File: 07.compiler/analyzer.go
package main

import (
	"regexp"
)

func NewAnalyzer(directoryPath string) *Analyzer {
	var tokenizedFiles []TokenizedFile

	parser := NewParser()
	tokenizer := NewTokenizer()
	compilationEngine := NewCompilationEngine(tokenizedFiles)

	return &Analyzer{
		parsedContent:     []ParsedContent{},
		directoryPath:     directoryPath,
		parser:            parser,
		tokenizer:         tokenizer,
		tokenizedFiles:    tokenizedFiles,
		compilationEngine: compilationEngine,
	}
}

func (a *Analyzer) Initialize() ([]ParsedContent, error) {
	var err error
	a.parsedContent, err = a.parser.ReadFilesFromDirectory(a.directoryPath)
	return a.parsedContent, err
}

func (a *Analyzer) PrepareContent() error {
	_, err := a.Initialize()
	if err != nil {
		return err
	}

	for _, content := range a.parsedContent {
		tokenizedLines := []Tokenized{}
		
		re := regexp.MustCompile(`"[^"]*"|\S+`)

		for _, line := range content.Content {
			tokens := re.FindAllString(line, -1)
			
			for _, token := range tokens {
				tokenized, err := a.tokenizer.Tokenize(token)
				if err != nil {
					return err
				}

				tokenizedLines = append(tokenizedLines, tokenized)
			}
		}

		a.tokenizedFiles = append(a.tokenizedFiles, TokenizedFile{
			Filename: content.Filename,
			Content:  tokenizedLines,
		})
	}

	return nil
}

func (a *Analyzer) Compile() error {
	err := a.PrepareContent()
	if err != nil {
		return err
	}

	a.compilationEngine.UpdateFiles(a.tokenizedFiles)
	return a.compilationEngine.Compile()
}

File: draft/compiler/analyzer.ts
import { CompilationEngine } from './compilation-engine';
import { Parser } from './parser';
import { Tokenizer } from './tokenizer';
import { ParsedContent, Tokenized, TokenizedFile } from './types';

export class Analyzer {
    parsedContent: Array<ParsedContent>;
    directoryPath: string;
    parser: Parser;
    tokenizer: Tokenizer;
    tokenizedFiles: TokenizedFile[];
    compilationEngine: CompilationEngine;

    constructor(diretoryPath: string) {
        this.parsedContent = [];
        this.directoryPath = diretoryPath;
        this.parser = new Parser();
        this.tokenizer = new Tokenizer();
        this.tokenizedFiles = [];
        this.compilationEngine = new CompilationEngine(this.tokenizedFiles);
    }

    public async initialize(): Promise<Array<ParsedContent>> {
        this.parsedContent = await this.parser.readFilesFromDirectory(
            this.directoryPath
        );
        return this.parsedContent;
    }

    public async prepareContent(): Promise<void> {
        await this.initialize();
        for (const content of this.parsedContent) {
            const tokenizedLines: Array<Tokenized> = [];
            for (const line of content.content) {
                const tokens = line.match(/"[^"]*"|\S+/g) || [];

                for (const token of tokens) {
                    const tokenized = this.tokenizer.tokenize(token);
                    if (typeof tokenized === 'string') {
                        return;
                    }

                    tokenizedLines.push(tokenized);
                }
            }
            this.tokenizedFiles.push({
                filename: content.filename,
                content: tokenizedLines,
            });
        }
    }

    public async compile(): Promise<void> {
        await this.prepareContent();
        this.compilationEngine.updateFiles(this.tokenizedFiles);
        this.compilationEngine.compile();
    }
}

const analyzer = new Analyzer(__dirname);

analyzer.compile();
